File: 백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque
import sys

input = sys.stdin.readline

n = int(input())
G = [list(map(int, input().split())) for _ in range(n)]

def can_go(x, y):
    return 0 <= x < n and 0 <= y < n

ans = 1
for h in range(1, 101):
    at_least_one_plus = False
    for i in range(n):
        for j in range(n):
            G[i][j] -= 1
            if G[i][j] > 0:
                at_least_one_plus = True
                
    if at_least_one_plus == False:
        break
            
    visited = [[False for _ in range(n)] for _ in range(n)]
    cnt = 0
    for i in range(n):
        for j in range(n):
            if G[i][j] > 0 and not visited[i][j]:
                q = deque()
                q.append((i, j))
                visited[i][j] = True
                cnt += 1
                while(q):
                    x, y  = q.popleft()
                    
                    for dx, dy in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                        nx, ny = x + dx, y + dy
                        
                        if can_go(nx, ny) and G[nx][ny] > 0 and not visited[nx][ny]:
                            q.append((nx, ny))
                            visited[nx][ny] = True
    ans = max(ans, cnt)
print(ans)
    

File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <iostream>
#include <deque>
#include <vector>

using namespace std;

int n;
int dx[4] = {0, 1, 0, -1};
int dy[4] = {1, 0, -1, 0};

bool in_range(int a, int b) {
    return 0 <= a && a < n && 0 <= b && b < n;
}

struct Node {
    int x;
    int y;
};

int main() {
    cin >> n;
    
    vector<vector<int>> G(n, vector<int>(n));
    int max_h = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> G[i][j];
            max_h = max(max_h, G[i][j]);
        }
    }
    
    int ans = 1;
    for (int h = 1; h < max_h; h++) {
        vector<vector<bool>> visited(n, vector<bool>(n, false));
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (G[i][j] > h && !visited[i][j]) {
                    cnt++;
                    deque<Node> q;
                    q.push_back({i, j});
                    visited[i][j] = true;
                    
                    while(!q.empty()) {
                        Node cur = q.front(); q.pop_front();
                        
                        int x = cur.x, y = cur.y;
                        
                        for (int dir = 0; dir < 4; dir++) {
                            int nx = x + dx[dir], ny = y + dy[dir];
                            
                            if (in_range(nx, ny) && G[nx][ny] > h && !visited[nx][ny]) {
                                q.push_back({nx, ny});
                                visited[nx][ny] = true;
                            }
                        }
                    }
                }
            }
        }
        ans = max(ans, cnt);
    }
    cout << ans;
    return 0;
}